### PR TITLE
Feature: Server-end progressive reader for http protocol

### PR DESCRIPTION
**Pull request by TangSiyang2001**
### What problem does this PR solve?

Issue Number: #2145

Problem Summary:

### What is changed and the side effects?

Changed:

1. Add a service-opt to enable server end progressive read. 
2. Check after all headers are completed, and determine whther reading request body in progressive way.
3. Users set progressive reader in user-defined service.

Side effects:
- Performance effects(性能影响):

- Breaking backward compatibility(向后兼容性): 

---
### Check List:
- Please make sure your changes are compilable(请确保你的更改可以通过编译).
- When providing us with a new feature, it is best to add related tests(如果你向我们增加一个新的功能, 请添加相关测试).
- Please follow [Contributor Covenant Code of Conduct](../../master/CODE_OF_CONDUCT.md).(请遵循贡献者准则).

### CODE REVIEW

File: src/brpc/details/http_message.h [hunk 1]
```diff
@@ -82,6 +82,10 @@ class HttpMessage {
 
     bool read_body_progressively() const { return _read_body_progressively; }
 
+    void set_read_progressively(bool read_body_progressively) {
```
Comment (wwbmmm) [inline src/brpc/details/http_message.h:85]: set_read_body_progressively

File: src/brpc/policy/http_rpc_protocol.cpp [hunk 2]
```diff
@@ -1545,6 +1549,26 @@ void ProcessHttpRequest(InputMessageBase *msg) {
     } else {
         // A http server, just keep content as it is.
         cntl->request_attachment().swap(req_body);
+        if (imsg_guard->read_body_progressively()) {
+            accessor.set_readable_progressive_attachment(imsg_guard.get());
+            const int sc = req_header.status_code();
```
Comment (wwbmmm) [inline src/brpc/policy/http_rpc_protocol.cpp:1554]: request header has status_code?

Comment (TangSiyang2001) [reply src/brpc/policy/http_rpc_protocol.cpp:1554]: Sorry, I borrowed some logic from `ProcessHttpResponse` and forgot to check the details, I'll modify it. lol.

File: src/brpc/policy/http_rpc_protocol.cpp [hunk 3]
```diff
@@ -1145,16 +1146,20 @@ ParseResult ParseHttpMessage(butil::IOBuf *source, Socket *socket,
...
             }
             return result;
-        } else if (socket->is_read_progressive() &&
-                   http_imsg->stage() >= HTTP_ON_HEADERS_COMPLETE) {
-            // header part of a progressively-read http message is complete,
-            // go on to ProcessHttpXXX w/o waiting for full body.
-            http_imsg->AddOneRefForStage2(); // released when body is fully read
-            return MakeMessage(http_imsg);
+        } else if (http_imsg->stage() >= HTTP_ON_HEADERS_COMPLETE) {
+            http_imsg->CheckProgressiveRead(arg, socket);
```
Comment (wwbmmm) [inline src/brpc/policy/http_rpc_protocol.cpp:1155]: Adding a method lookup in hot path, maybe affect performance

Comment (TangSiyang2001) [reply src/brpc/policy/http_rpc_protocol.cpp:1155]: OK, I'll make `CheckProgressiveRead` inline.

Comment (wwbmmm) [reply src/brpc/policy/http_rpc_protocol.cpp:1155]: Not inline, I mean the FindMethodPropertyByURI call inside CheckProgressiveRead.

Comment (TangSiyang2001) [reply src/brpc/policy/http_rpc_protocol.cpp:1155]: Oh I got u. However, one of the keypoints of this feature is to get the corresponding service opt `enable_progressive_read` after header completed and determine the read body behavior. Is there a better way to get this opt without  calling `FindMethodPropertyByURI` to fetch it inside the method property?  Appreciate it.

Comment (wwbmmm) [reply src/brpc/policy/http_rpc_protocol.cpp:1155]: Maybe we can add a flag `has_progressive_read_method` to Server
When Server Start, check all MethodProperty, if any one has progressive_read option, set `has_progressive_read_method` to true
When parsing message, check `has_progressive_read_method`, if true, then call FindMethodPropertyByURI

Comment (TangSiyang2001) [reply src/brpc/policy/http_rpc_protocol.cpp:1155]: Great idea! Thank u so much.

**Comment by TangSiyang2001** [comment]
@wwbmmm PTAL, thx.

File: test/brpc_http_rpc_protocol_unittest.cpp [hunk 4]
```diff
@@ -1137,7 +1137,7 @@ TEST_F(HttpTest, server_end_read_short_body_progressively) {
 }
 
 TEST_F(HttpTest, server_end_read_failed) {
-    const int port = 8923;
+     const int port = 8923;
```
Comment (wwbmmm) [inline test/brpc_http_rpc_protocol_unittest.cpp:1140]: 缩进不对

File: src/brpc/server.h [hunk 5]
```diff
@@ -560,10 +565,14 @@ class Server {
     int Concurrency() const {
         return butil::subtle::NoBarrier_Load(&_concurrency);
     };
-
+  
     // Returns true if accept request, reject request otherwise.
     bool AcceptRequest(Controller* cntl) const;
 
+    bool HasProgressiveReadMethod() const {
```
Comment (wwbmmm) [inline src/brpc/server.h:572]: 简单的getter/setter函数，用下划线分隔的全小写命名方式

File: test/brpc_http_rpc_protocol_unittest.cpp (hunk 4 above)
Comment (TangSiyang2001) [reply test/brpc_http_rpc_protocol_unittest.cpp]: modified

File: src/brpc/server.h (hunk 5 above)
Comment (TangSiyang2001) [reply src/brpc/server.h]: modified

**Comment by wwbmmm** [comment]
LGTM

**Comment by TangSiyang2001** [comment]
> LGTM

@wwbmmm Could this pr be merged?

**Pull request merged**